### PR TITLE
8315412 [lworld] Preparing code for lw5

### PR DESCRIPTION
**Pull request by fparain**
Quite a big patch, but it is mostly made of renaming and mechanical replacement without really changing the logic of the code.

The patch includes changes to make the VM code less dependent on Q-descriptors by encoding the presence of null-free inline types fields in FieldInfo.

The patch also includes a lot of renaming, in an effort to have uniformed naming of flat fields and flat arrays across the VM. The renaming has not been applied to the heap dumper and C2. They will be addressed in follow up patches.

Tested with Mach5, tiers 1 to 3.

Thank you

Fred

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8315412](https://bugs.openjdk.org/browse/JDK-8315412): [lworld] Preparing code for lw5 (**Enhancement** - P4)

### Reviewers
 * [Lois Foltan](https://openjdk.org/census#lfoltan) (@lfoltan - Committer) ⚠️ Review applies to [9fa21a50](https://git.openjdk.org/valhalla/pull/922/files/9fa21a50700bf6b11b486d64db3582d9fda68986)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - Committer) ⚠️ Review applies to [80e6f6c7](https://git.openjdk.org/valhalla/pull/922/files/80e6f6c76c80bc368c3f83a7f53349a6424982f7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/922/head:pull/922` \
`$ git checkout pull/922`

Update a local copy of the PR: \
`$ git checkout pull/922` \
`$ git pull https://git.openjdk.org/valhalla.git pull/922/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 922`

View PR using the GUI difftool: \
`$ git pr show -t 922`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/922.diff">https://git.openjdk.org/valhalla/pull/922.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/922#issuecomment-1699850363)

### CODE REVIEW

File: src/hotspot/share/classfile/fieldLayoutBuilder.cpp [hunk 1]
```diff
@@ -692,7 +694,7 @@ void FieldLayoutBuilder::regular_field_sorting() {
  *     constraining alignment, this value is then used as the alignment
  *     constraint when flattening this inline type into another container
  *   - field flattening decisions are taken in this method (those decisions are
- *     currently only based in the size of the fields to be inlined, the size
+ *     currently only based in the size of the fields to be flatten, the size
```
Comment (lfoltan) [inline src/hotspot/share/classfile/fieldLayoutBuilder.cpp:697]: flatten -> flattened

File: src/hotspot/share/oops/cpCache.cpp [hunk 2]
```diff
@@ -123,17 +123,17 @@ void ConstantPoolCacheEntry::set_field(Bytecodes::Code get_code,
...
   set_f2(field_offset);
   assert((field_index & field_index_mask) == field_index,
          "field index does not fit in low flag bits");
-  assert(!is_inlined || is_null_free_inline_type, "Sanity check");
+  assert(!is_flat || is_null_free_inline_type, "Sanity check");
   set_field_flags(field_type,
                   ((is_volatile ? 1 : 0) << is_volatile_shift) |
                   ((is_final    ? 1 : 0) << is_final_shift) |
-                  ((is_inlined  ? 1 : 0) << is_inlined_shift) |
+                  ((is_flat  ? 1 : 0) << is_flat_shift) |
```
Comment (lfoltan) [inline src/hotspot/share/oops/cpCache.cpp:136]: Minor nit, line up ternary expression with lines above

File: src/hotspot/share/ci/ciInstanceKlass.cpp [hunk 3]
```diff
@@ -770,57 +770,59 @@ void StaticFieldPrinter::do_field_helper(fieldDescriptor* fd, oop mirror, bool f
       _out->print(INT64_FORMAT, *(int64_t*)&d);
       break;
     }
+    case T_PRIMITIVE_OBJECT: // fall-through
     case T_ARRAY:  // fall-through
```
Comment (lfoltan) [inline src/hotspot/share/ci/ciInstanceKlass.cpp:774]: Minor nit: do both line 773 & 774 need "fall-through" comment?

**Review by lfoltan — APPROVED** [review]
Good clean up, some very minor comments, overall looks good.

File: src/hotspot/share/classfile/fieldLayoutBuilder.cpp (hunk 1 above)
Comment (fparain) [reply src/hotspot/share/classfile/fieldLayoutBuilder.cpp]: Fixed

File: src/hotspot/share/ci/ciInstanceKlass.cpp (hunk 3 above)
Comment (fparain) [reply src/hotspot/share/ci/ciInstanceKlass.cpp:774]: The comment in the T_ARRAY case will remain even after the removal of Q-descriptor.
The comment in the T_PRIMITIVE_OBJECT case makes it much easier to see (with grep) that this case is simply going to go away after the removal of Q-descriptors.

File: src/hotspot/share/oops/cpCache.cpp (hunk 2 above)
Comment (fparain) [reply src/hotspot/share/oops/cpCache.cpp]: Fixed

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 4]
```diff
@@ -2940,25 +2940,25 @@ void MacroAssembler::test_oop_prototype_bit(Register oop, Register temp_reg, int
   jcc((jmp_set) ? Assembler::notZero : Assembler::zero, jmp_label);
 }
 
-void MacroAssembler::test_flattened_array_oop(Register oop, Register temp_reg,
-                                              Label&is_flattened_array) {
+void MacroAssembler::test_flat_array_oop(Register oop, Register temp_reg,
+                                              Label& is_flat_array) {
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:2944]: Indentation needs to be fixed.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 5]
```diff
@@ -2940,25 +2940,25 @@ void MacroAssembler::test_oop_prototype_bit(Register oop, Register temp_reg, int
...
   movl(temp_reg, Address(temp_reg, Klass::layout_helper_offset()));
-  test_flattened_array_layout(temp_reg, is_flattened_array);
+  test_flat_array_layout(temp_reg, is_flat_array);
 #endif
 }
 
-void MacroAssembler::test_non_flattened_array_oop(Register oop, Register temp_reg,
-                                                  Label&is_non_flattened_array) {
+void MacroAssembler::test_non_flat_array_oop(Register oop, Register temp_reg,
+                                                  Label& is_non_flat_array) {
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:2955]: Indentation needs to be fixed.

File: src/hotspot/cpu/x86/interp_masm_x86.hpp [hunk 6]
```diff
@@ -240,16 +240,16 @@ class InterpreterMacroAssembler: public MacroAssembler {
...
 
-  // Allocate value buffer in "obj" and read in flattened element at the given index
+  // Allocate value buffer in "obj" and read in flat element at the given index
   // NOTES:
   //   - Return via "obj" must be rax
   //   - kills all given regs
   //   - 32 bits: kills rdi and rsi
-  void read_flattened_element(Register array, Register index,
+  void read_flat_element(Register array, Register index,
                               Register t1, Register t2,
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/interp_masm_x86.hpp:253]: Indentation needs to be fixed.

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp [hunk 7]
```diff
@@ -285,7 +285,7 @@ void InterpreterMacroAssembler::allocate_instance(Register klass, Register new_o
   }
 }
 
-void InterpreterMacroAssembler::read_inlined_field(Register holder_klass,
+void InterpreterMacroAssembler::read_flat_field(Register holder_klass,
                                                    Register field_index, Register field_offset,
```
Comment (TobiHartmann) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:289]: Indentation needs to be fixed.

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.hpp [hunk 8]
```diff
@@ -154,15 +154,15 @@ class InterpreterMacroAssembler: public MacroAssembler {
   //   - input holder object via "obj", which must be r0,
   //     will return new instance via the same reg
   //   - assumes holder_klass and valueKlass field klass have both been resolved
-  void read_inlined_field(Register holder_klass,
+  void read_flat_field(Register holder_klass,
                           Register field_index, Register field_offset,
```
Comment (TobiHartmann) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.hpp:158]: Indentation needs to be fixed.

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.hpp [hunk 9]
```diff
@@ -154,15 +154,15 @@ class InterpreterMacroAssembler: public MacroAssembler {
...
 
-  // Allocate value buffer in "obj" and read in flattened element at the given index
+  // Allocate value buffer in "obj" and read in flat element at the given index
   // NOTES:
   //   - Return via "obj" must be r0
   //   - kills all given regs
-  void read_flattened_element(Register array, Register index,
+  void read_flat_element(Register array, Register index,
                               Register t1, Register t2,
                               Register obj = r0);
```
Comment (TobiHartmann) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.hpp:167]: Indentation needs to be fixed.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 10]
```diff
@@ -1246,7 +1246,7 @@ void InterpreterMacroAssembler::allocate_instance(Register klass, Register new_o
 }
 
 
-void InterpreterMacroAssembler::read_inlined_field(Register holder_klass,
+void InterpreterMacroAssembler::read_flat_field(Register holder_klass,
                                                      Register field_index, Register field_offset,
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:1250]: Indentation needs to be fixed.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 11]
```diff
@@ -1286,13 +1286,13 @@ void InterpreterMacroAssembler::read_inlined_field(Register holder_klass,
...
+  call_VM(obj, CAST_FROM_FN_PTR(address, InterpreterRuntime::read_flat_field),
           obj, field_index, holder_klass);
 
   bind(done);
 }
 
-void InterpreterMacroAssembler::read_flattened_element(Register array, Register index,
+void InterpreterMacroAssembler::read_flat_element(Register array, Register index,
                                                        Register t1, Register t2,
                                                        Register obj) {
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:1297]: Indentation needs to be fixed.

File: src/hotspot/share/c1/c1_LIRGenerator.cpp [hunk 12]
```diff
@@ -2352,19 +2352,19 @@ void LIRGenerator::do_LoadIndexed(LoadIndexed* x) {
...
   Value element;
   if (x->vt() != nullptr) {
-    assert(x->array()->is_loaded_flattened_array(), "must be");
+    assert(x->array()->is_loaded_flat_array(), "must be");
     // Find the destination address (of the NewInlineTypeInstance).
     LIRItem obj_item(x->vt(), this);
 
-    access_flattened_array(true, array, index, obj_item,
+    access_flat_array(true, array, index, obj_item,
                            x->delayed() == nullptr ? 0 : x->delayed()->field(),
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_LIRGenerator.cpp:2360]: Indentation needs to be fixed.

File: src/hotspot/share/c1/c1_Runtime1.hpp [hunk 13]
```diff
@@ -54,8 +54,8 @@ class StubAssembler;
   stub(new_object_array)             \
   stub(new_flat_array)               \
   stub(new_multi_array)              \
-  stub(load_flattened_array)         \
-  stub(store_flattened_array)        \
+  stub(load_flat_array)         \
+  stub(store_flat_array)        \
```
Comment (TobiHartmann) [inline src/hotspot/share/c1/c1_Runtime1.hpp:58]: Indentation needs to be fixed.

File: src/hotspot/share/ci/ciInstanceKlass.cpp [hunk 14]
```diff
@@ -770,57 +770,59 @@ void StaticFieldPrinter::do_field_helper(fieldDescriptor* fd, oop mirror, bool f
...
-        obj = mirror->obj_field_acquire(fd->offset());
+        // handling of null free inline type
+        ResetNoHandleMark rnhm;
+        Thread* THREAD = Thread::current();
+        SignatureStream ss(fd->signature(), false);
+        Symbol* name = ss.as_symbol();
+        assert(!HAS_PENDING_EXCEPTION, "can resolve klass?");
+        InstanceKlass* holder = fd->field_holder();
+        InstanceKlass* k = SystemDictionary::find_instance_klass(THREAD, name,
+                                                                Handle(THREAD, holder->class_loader()),
```
Comment (TobiHartmann) [inline src/hotspot/share/ci/ciInstanceKlass.cpp:811]: Indentation needs to be fixed.

File: src/hotspot/share/interpreter/interpreterRuntime.cpp [hunk 15]
```diff
@@ -362,15 +362,15 @@ JRT_ENTRY(int, InterpreterRuntime::withfield(JavaThread* current, ConstantPoolCa
     case atos:
       {
         if (cpe->is_null_free_inline_type())  {
-          if (!cpe->is_inlined()) {
+          if (!cpe->is_flat()) {
               if (ref_h() == nullptr) {
```
Comment (TobiHartmann) [inline src/hotspot/share/interpreter/interpreterRuntime.cpp:366]: Indentation needs to be fixed.

**Review by TobiHartmann — APPROVED** [review]
Nice cleanup! I only found some indentation issues.

**Comment by fparain** [comment]
Lois, Tobias,
thank you for your reviews, comments and indentations have been fixed.

Fred

**Pull request closed without merging**